Lib/_pylong.py

```diff
@@ -210,6 +210,91 @@ def inner(a, b):
     w5pow = compute_powers(len(s), 5, DIGLIM)
     return inner(0, len(s))
 
+# Asymptotically faster version, using the C decimal module. See
+# comments at the end of the file. This uses decimal arithmetic to
+# convert from base 10 to base 256. The latter is just a string of
+# bytes, which CPython can convert very efficiently to a Python int.
+
+import math
+_LOG_10_BASE_256 = math.log(10, 256)
+del  math
+
+def _dec_str_to_int_inner(s):
+    BYTELIM = 512
+    D = decimal.Decimal
+    result = bytearray()
+
+    def inner(n, w):
+        if w <= BYTELIM:
+            # XXX Stefan Pochmann discovered that, for 1024-bit ints,
+            # `int(Decimal)` took 2.5x longer than `int(str(Decimal))`.
+            # So simplify this code to the former if/when that gets
+            # repaired.
+            result.extend(int(str(n)).to_bytes(w)) # big-endian default
+            return
+        w2 = w >> 1
+        if 0:
+            # This is maximally clear, but "too slow". `decimal`
+            # division is asymptotically fast, but we have no way to
+            # tell it to reuse the high-precision reciprocal it computes
+            # for pow256[w2], so it has to recompute it over & over &
+            # over again :-(
+            hi, lo = divmod(n, pow256[w2][0])
+        else:
+            p256, recip = pow256[w2]
+            # The integer part will have about half the digits of n. So
+            # only need that much precision, plus some guard digits.
+            ctx.prec = (n.adjusted() >> 1) + 8
```

Comment:
Did you try this with a smaller number of guard digits or without them at all? Does it still return the correct result (but slower)?

Reply:
I have since replaced that code with a better method to pick the working precision, and with a much better explanation for "why" it's right.

Yes, I tried other values for guard digits. A lower bound on the quotient is always computed. The difference is in how many times the correction loop may need to go around. At 8, I've never seen it go around more than once, and even that for only 1 in about 100 thousand randomized inputs. It could be made smaller without harming speed, but it wouldn't measurably help speed either - 8 digits more or less is small in this context.

Reply:
I worry that for very large numbers it may require more than 1 correction, and that `assert count < 10` can fail. How smaller values for guard digits (1 or 2) affect the number of corrections? Is it always not more than 1 or 2? If it significantly grows, I would remove the `assert`.

Reply:
I use `assert` for conditions I _believe_ are impossible to fail. But it's always the case that I may be wrong.

The number of digits shouldn't matter. This is about what can happen at the _last_ retained digit. With finite precision, it's always possible for the quotient to be too low by 1, but that can be made ever less likely by increasing the number of guard digits (so accumulated errors show up directly at _their_ low end instead).

Various analysis & coding errors while developing this convinced me that when the code is wrong, there's effectively no limit on how often the correction loop can go around. Better to raise an assert failure then than for the user's program to appear to hang using 100% CPU.

But, ya, with 8 guard digits I _believe_ 1 is the largest correction needed (no matter how large the input). I expect that with 7 guards too, or even 6. But below that, it gets fuzzier. As the newer code comments explain, it's the log10s of `n` and `p256` that matter to where the last integer-part digit is, and for both speed and simplicity I'm settling for `.adjusted()` as a very crude approximation to what those logs are. Guard digits are very effective at preventing bad consequences from incomplete analysis..

Indeed, I would have far less confidence in a purported "proof" of how badly the various approximations could end up hurting in the absence of guards.

Reply:
Nope! Changed my mind again - 1 is the largest correction ever needed. But proof is involved. Short course: if `prec` is the decimal precision in effect, and we're rounding down, the result of an operation is exactly equal to the infinitely precise result times `1 - e` for some real `e` with `0 <= e < 10**(1-prec`). We have 3 operations: chopping `n` back to `prec` digits, likewise for `1/256**w2`, and also for their product.

So the computed product is exactly equal to the true product times `(1-e1)*(1-e2)*(1-e3)`; since the `e`s are all very small, an excellent approximation to the second factor is `1-(e1+e2+e3)` (the 2nd and 3rd order terms in the expanded product are too tiny to matter). If they're all as large as possible, that's `1 - 3*10**(1-prec)`. This, BTW, is all bog-standard FP error analysis.

Skipping details, that implies the computed product is within 1 of the true product provided `prec >= log10(true_product) + 1.47712125`.

`n.adjusted() - p256.adjusted()` is s crude integer approximation to `log10(true_product)` - but `prec` is necessarily an int too, so adding another guard digit or two is always enough to compensate for that. Again skipping details.

Also skipping why cutting the reciprocal to `p256.adjusted() + GUARD` digits to begin with is good enough. The precondition `n < 256**w` is needed to establish that the true product can't be too large for the reciprocal approximation to be too narrow.

And skipping other assorted details. I don't intend to change the code again anyway, or to write up such an involved proof in detail, but I'll probably add some high-level comments, akin to this one.